src/jit/lsra.cpp

```diff
@@ -4857,15 +4901,40 @@ LinearScan::tryAllocateFreeReg(Interval *currentInterval, RefPosition *refPositi
 //                  and that can be spilled.
 //
 // Arguments:
-//    current:         The interval for the current allocation
-//    refPosition:     The RefPosition of the current Interval for which a register is being allocated
+//    current               The interval for the current allocation
+//    refPosition           The RefPosition of the current Interval for which a register is being allocated
+//    allocationOptional    If true, a reg may not be allocated if all other ref positions currently
+//                          occupying registers are more important than the 'refPosition'.
 //
 // Return Value:
 //    The regNumber allocated to the RefPositon.  Returns REG_NA if no free register is found.
-
+//
+// Note:  Currently this routine uses weight and farthest distance of next reference
+// to select a ref position for spilling.  
+// a) if allocationOptional = false
+//        The ref position chosen for spilling will be the lowest weight
+//        of all and if there is is more than one ref position with the
+//        same lowest weight, among them choses the one with farthest
+//        distance to its next reference.
+// 
+// b) if allocationOptional = true
+//        The ref position chosen for spilling will not only be lowest weight
+//        of all but also has a weight lower than 'refPosition'.  If there is
+//        no such ref position, reg will not be allocated.
 regNumber
-LinearScan::allocateBusyReg(Interval *current, RefPosition *refPosition)
+LinearScan::allocateBusyReg(Interval* current,
+                            RefPosition* refPosition,
+                            bool allocationOptional)
 {
```

Comment:
Right now RequiresRegister() is used in many places to mean a ref position that needs to be allocated a reg (e.g. refTypeDef and RefTypeUse)

Whereas RequiresRegisterOptionally() applies only to RefTypeUse.

RegOptional means that it doesn't require or  be allocated a register, but it just means it is ok even if no reg is allocated.

I considered renaming RequiresRegister() but couldn't come up with a suitable name.  Finally I have started to look at these routines as follows

RequiresRegister() - These are the ref positions to which reg are allocated.
RequiresRegisterOptionally() -  This indicates that reg allocation is optional, but it can also be considered for reg allocation.

I know, this is still a little confusing.  I am open to renaming RequiresRegisterOptionally() to IsRegOptional().

Do you have any suggestion here?